viz/pkg/healthcheck/healthcheck.go

```diff
@@ -192,9 +182,123 @@ func (hc *HealthChecker) vizCategory() *healthcheck.Category {
 			WithRetryDeadline(hc.RetryDeadline).
 			WithCheckRPC(func(ctx context.Context) (*healthcheckPb.SelfCheckResponse, error) {
 				return hc.vizAPIClient.SelfCheck(ctx, &healthcheckPb.SelfCheckRequest{})
-			}))
+			}),
+	}, true)
+}
+
+// VizDataPlaneCategory returns a healthcheck.Category containing checkers
+// to verify the data-plane metrics in prometheus and the tap injection
+func (hc *HealthChecker) VizDataPlaneCategory() healthcheck.Category {
+
+	return *healthcheck.NewCategory(LinkerdVizExtensionDataPlaneCheck, []healthcheck.Checker{
+		*healthcheck.NewChecker("data plane namespace exists").
+			WithHintAnchor("l5d-data-plane-exists").
+			Fatal().
+			WithCheck(func(ctx context.Context) error {
+				if hc.DataPlaneNamespace == "" {
+					// when checking proxies in all namespaces, this check is a no-op
+					return nil
+				}
+				return hc.CheckNamespace(ctx, hc.DataPlaneNamespace, true)
+			}),
+		*healthcheck.NewChecker("data plane proxy metrics are present in Prometheus").
+			WithHintAnchor("l5d-data-plane-prom").
+			Warning().
+			WithRetryDeadline(hc.RetryDeadline).
+			WithCheck(func(ctx context.Context) (err error) {
+				pods, err := hc.getDataPlanePodsFromVizAPI(ctx)
+				if err != nil {
+					return err
+				}
+
+				// TODO: Check if prometheus is present
+
+				return validateDataPlanePodReporting(pods)
+			}),
+		*healthcheck.NewChecker("data-plane pods have the correct tap configuration").
+			WithHintAnchor("l5d-viz-data-plane-tap").
+			Warning().
+			WithCheck(func(ctx context.Context) error {
+				pods, err := hc.GetDataPlanePods(ctx)
+				if err != nil {
+					return err
+				}
+
+				return hc.checkForTapConfiguration(ctx, pods)
+			}),
+	}, true)
+}
+
+func (hc *HealthChecker) getDataPlanePodsFromVizAPI(ctx context.Context) ([]*pb.Pod, error) {
+
+	req := &pb.ListPodsRequest{}
+	if hc.DataPlaneNamespace != "" {
+		req.Selector = &pb.ResourceSelection{
+			Resource: &pb.Resource{
+				Namespace: hc.DataPlaneNamespace,
+			},
+		}
+	}
+
+	resp, err := hc.VizAPIClient().ListPods(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	pods := make([]*pb.Pod, 0)
+	for _, pod := range resp.GetPods() {
+		if pod.ControllerNamespace == hc.ControlPlaneNamespace {
+			pods = append(pods, pod)
+		}
+	}
+
+	return pods, nil
+}
+
+// checkForTapConfiguration checks if the tap annotation is present
+// only for the pods with tap enabled
+func (hc *HealthChecker) checkForTapConfiguration(ctx context.Context, pods []corev1.Pod) error {
+	var podsWithoutTap []string
+	for _, pod := range pods {
+		ns, err := hc.KubeAPIClient().CoreV1().Namespaces().Get(ctx, pod.Namespace, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		// Check if Tap is disabled
+		if !k8s.IsTapDisabled(pod) && !k8s.IsTapDisabled(ns) {
+			// Check for tap-injector annotation
+			if _, contains := pod.GetAnnotations()[vizLabels.VizTapEnabled]; !contains {
+				podsWithoutTap = append(podsWithoutTap, fmt.Sprintf("* %s", pod.Name))
+			}
+		}
```

Comment:
Now that we can go off the annotation, I think this can be simplified a little more.

The tap injector will not add the `tap-enabled` annotation to a pod if tap is disabled via annotation on the pod or the namespace. Therefore, we don't need to check for the `disable-tap` annotation on the namespace or the pod. We should be okay to skip getting the namespace and only check the the pod has the annotation.

Additionally, there is already a `vizLabels.IsTapEnabled` function we can use here. It also checks the value of the annotation is actually `true`.

Reply:
Talked off GitHub about it, but never the part about removing the tap disabled check. We'll need that since we'd then issue warnings on pods that don't have the `tap-enabled` annotation, but do have the `disable-tap` annotation.